source/common/buffer/buffer_impl.h

```diff
@@ -297,11 +310,30 @@ class Slice {
    * Call all drain trackers associated with the slice, then clear
    * the drain tracker list.
    */
-  void callAndClearDrainTrackers() {
+  void callAndClearDrainTrackersAndCharges() {
     for (const auto& drain_tracker : drain_trackers_) {
       drain_tracker();
     }
     drain_trackers_.clear();
+
+    if (account_) {
+      account_->credit(capacity_);
+      account_.reset();
+    }
+  }
+
+  /**
+   * Charges the provided account for the resources if these conditions hold:
+   * - we're not already charging for this slice
+   * - the given account is non-null
+   * - the slice owns backing memory
+   */
+  void maybeChargeAccount(const BufferMemoryAccountSharedPtr& account) {
+    if (account_ != nullptr || storage_ == nullptr || !account) {
```

Comment:
nit: `account == nullptr` instead of `!account`. You compare another shared_ptr to `nullptr` on the same line, so keep the formatting of comparisons the same.

Reply:
Done.